external-crates/move/crates/move-compiler/src/typing/syntax_methods.rs

```diff
@@ -0,0 +1,234 @@
+
+    let prev_constraints = std::mem::take(&mut context.constraints);
+
+    let mut valid = true;
+
+    let mut subst = core::Subst::empty();
+
+    let index_ty = core::make_function_type(context, index_ann_loc, index_module, index_fn, None);
+    let index_mut_ty =
+        core::make_function_type(context, index_ann_loc, index_mut_module, index_mut_fn, None);
+
+    for (ndx, (index_tyarg, index_mut_tyarg)) in index_ty
+        .ty_args
+        .iter()
+        .zip(index_mut_ty.ty_args.iter())
+        .enumerate()
+    {
+        if let Ok((subst_, _)) = core::join(subst.clone(), index_tyarg, index_mut_tyarg) {
+            subst = subst_;
+        } else {
+            context
+                .env
+                .add_diag(ice!((index.loc, "Failed in validating syntax methods")));
+            valid = false;
+        }
+        // This essentially fakes skolomeziation so that we can fail when the types aren't the same.
+        let id = N::TParamID::next();
+        let user_specified_name = sp(index_tyarg.loc, format!("_{}", ndx).into());
+        let tparam = N::TParam {
+            id,
+            user_specified_name,
+            abilities: AbilitySet::all(index_ann_loc),
+        };
+        if let Ok((subst_, _)) = core::join(
+            subst.clone(),
+            index_tyarg,
+            &sp(index_tyarg.loc, N::Type_::Param(tparam)),
+        ) {
+            subst = subst_;
+        } else {
+            context
+                .env
+                .add_diag(ice!((index.loc, "Failed in validating syntax methods")));
+            valid = false;
+        }
+    }
+
+    fn ty_str(ty: &N::Type) -> String {
+        core::error_format(ty, &core::Subst::empty())
+    }
+
+    fn ty_str_(ty: &N::Type_) -> String {
+        core::error_format_(ty, &core::Subst::empty())
+    }
+
+    if let Ok((subst_, _)) = core::subtype(
+        subst.clone(),
+        &index_mut_ty.params[0].1,
+        &index_ty.params[0].1,
+    ) {
+        subst = subst_;
+    } else {
+        let (_, _, index_type) = &index_finfo.signature.parameters[0];
+        let (_, _, mut_type) = &mut_finfo.signature.parameters[0];
+        // This case shouldn't really be reachable, but we might as well provide an error.
+        let index_msg = format!(
+            "This index function subject has type {}",
+            ty_str(index_type)
+        );
+        let mut_msg = format!(
+            "This mutable index function subject has type {}",
+            ty_str(mut_type)
+        );
+        let mut diag = diag!(
+            TypeSafety::IncompatibleSyntaxMethods,
+            (index_type.loc, index_msg),
+            (mut_type.loc, mut_msg)
+        );
+        diag.add_note(
+            "These functions must take the same subject type, differing only by mutability.",
+        );
+        context.env.add_diag(diag);
+        valid = false;
+    }
+
+    for (ndx, ((_, index_param), (_, index_mut_param))) in index_ty.params[1..]
```

Comment:
Won't this panic on an empty index?

Reply:
e.g. something like `#[syntax(index)] fun borrow(self: &Option<T>): &T`

Reply:
This does not, since Rust will just return an empty slice here. Even so, the iterator rewrite got rid of this code.